modules/urlsigning-service-impl/src/main/java/org/opencastproject/security/urlsigning/provider/impl/AbstractUrlSigningProvider.java

```diff
@@ -150,121 +109,113 @@ public void addKeyEntry(String keyId, String url, String key, String organizatio
-    String keyId = null;
-    String url = null;
-    String organization = null;
+    // Collect configuration in a new map so we don't partially override the old one in case of error
+    Map<String, KeyEntry> keys = new HashMap<>();
 
-    int i = 1;
-    while (true) {
-      // Create the configuration prefixes
-      key = new StringBuilder(KEY_PREFIX).append(".").append(i).toString();
-      keyId = new StringBuilder(ID_PREFIX).append(".").append(i).toString();
-      url = new StringBuilder(URL_PREFIX).append(".").append(i).toString();
-      organization = new StringBuilder(ORGANIZATION_PREFIX).append(".").append(i).toString();
-      getLogger().debug("Looking for configuration of {}, {}, {} and {}", key, keyId, url, organization);
-      // Read the key, keyId and url values
-      String keyValue = StringUtils.trimToNull((String) properties.get(key));
-      String keyIdValue = StringUtils.trimToNull((String) properties.get(keyId));
-      String urlValue = StringUtils.trimToNull((String) properties.get(url));
-      String organizationValue = StringUtils.trimToNull((String) properties.get(organization));
-
-      // Has the url signing provider been fully configured
-      // Note: organization is optional
-      if (keyValue == null || keyIdValue == null || urlValue == null) {
-        getLogger().debug(
-                "Unable to configure key with id '{}' and url matcher '{}' because the id, key or url is missing. Stopping to look for new keys.",
-                keyIdValue, urlValue);
-        break;
-      }
+    Enumeration<String> propertyKeys = properties.keys();
+    while (propertyKeys.hasMoreElements()) {
+      String propertyKey = propertyKeys.nextElement();
 
-      if (organizationValue == null) {
-        // organization is optimal. Set it to default
-        organizationValue = ANY_ORGANIZATION;
-        getLogger().debug("No organization set for key id {}, using default organization {}", keyIdValue,
-              organizationValue);
+      String[] parts = Arrays.stream(propertyKey.split("\\."))
+              .map(String::trim)
+              .toArray(String[]::new);
+      if (parts.length != 2) {
+        throw new ConfigurationException(propertyKey, "wrong property key format");
       }
 
-      // Store the key
-      try {
-        addKeyEntry(keyIdValue, urlValue, keyValue, organizationValue);
-        getLogger().info("{} will handle uris that start with '{}' with the key id '{}'", toString(), urlValue,
-                keyIdValue);
-      } catch (IllegalStateException e) {
-        throw new ConfigurationException(urlValue, e.getMessage());
+      String id = parts[0];
+      KeyEntry currentKeyEntry = keys.computeIfAbsent(id, __ -> new KeyEntry());
+
+      String attribute = parts[1];
+      String propertyValue = Objects.toString(properties.get(propertyKey), "").trim();
+      switch (attribute) {
+        case ORGANIZATION:
+          currentKeyEntry.organization = propertyValue;
```

Comment:
If `propertyValue` is the empty string here, is this a valid thing to do?

Reply:
Looking at the old code again, none of the property values were allowed to be `null` or empty. Somehow I missed adopting that behavior. Will fix.

Reply:
Note that this is resolved [here](https://github.com/opencast/opencast/pull/731/files#diff-34839f84f847d48880db263f855ba1c9R155).